segmentstore/server/src/test/java/io/pravega/segmentstore/server/containers/DebugStreamSegmentContainerTests.java

```diff
@@ -247,6 +251,85 @@ public void testEndToEnd() throws Exception {
+        @Cleanup
+        val s = new AsyncStorageWrapper(new RollingStorage(baseStorage, new SegmentRollingPolicy(1)), executorService());
+        s.initialize(1);
+        log.info("Created a storage instance");
+
+        String sourceSegmentName = "segment-" + RANDOM.nextInt();
+        String targetSegmentName = "segment-" + RANDOM.nextInt();
+
+        // Create source segment
+        s.create(sourceSegmentName, TIMEOUT).join();
+        val handle = s.openWrite(sourceSegmentName).join();
+
+        // do some writing
+        ByteArrayOutputStream writeStream = new ByteArrayOutputStream();
+        long offset = 0;
+        for (int j = 0; j < APPENDS_PER_SEGMENT; j++) {
+            byte[] writeData = populate(APPEND_FORMAT.length());
+
+            val dataStream = new ByteArrayInputStream(writeData);
+            s.write(handle, offset, dataStream, writeData.length, TIMEOUT).join();
+            writeStream.write(writeData);
+            offset += writeData.length;
+        }
+
+        // create a debug segment container instance
+        @Cleanup
+        TestContext context = createContext(executorService());
+        OperationLogFactory localDurableLogFactory = new DurableLogFactory(DEFAULT_DURABLE_LOG_CONFIG, context.dataLogFactory,
+                executorService());
+
+        @Cleanup
+        MetadataCleanupContainer localContainer = new MetadataCleanupContainer(0, CONTAINER_CONFIG, localDurableLogFactory,
+                context.readIndexFactory, context.attributeIndexFactory, context.writerFactory, context.storageFactory,
+                context.getDefaultExtensions(), executorService());
+        Services.startAsync(localContainer, executorService()).join();
+
+        // copy segment
+        localContainer.copySegment(s, sourceSegmentName, targetSegmentName, executorService()).join();
+
+        // new segment should exist
+        Assert.assertTrue("Unexpected result for existing segment (no files).", s.exists(sourceSegmentName, null).join());
+        // Old segment should exist
+        Assert.assertTrue("Unexpected result for missing segment (no files).", s.exists(targetSegmentName, null).join());
+
+        // Do some reading.
+        val readHandle = s.openRead(targetSegmentName).join();
+        byte[] expectedData = writeStream.toByteArray();
+
+        for (offset = 0; offset < expectedData.length / 2; offset++) {
+            int length = (int) (expectedData.length - 2 * offset);
+            byte[] readBuffer = new byte[length];
+            int bytesRead = s.read(readHandle, offset, readBuffer, 0, readBuffer.length, TIMEOUT).join();
+            Assert.assertEquals(String.format("Unexpected number of bytes read from offset %d.", offset),
+                    length, bytesRead);
+            AssertExtensions.assertArrayEquals(String.format("Unexpected read result from offset %d.", offset),
+                    expectedData, (int) offset, readBuffer, 0, bytesRead);
+        }
+    }
```

Comment:
Not sure why this needs to be in loop

Reply:
Removed data. Wrote 10 MB data, and reading all of it at once and verifying.